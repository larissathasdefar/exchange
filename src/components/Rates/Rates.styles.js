import styled from 'styled-components'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import Button from '@material-ui/core/Button'
import NativeSelect from '@material-ui/core/NativeSelect'
import TableRow from '@material-ui/core/TableRow'
import IconButton from '@material-ui/core/IconButton'

export const Container = styled(Paper)`
  padding: 20px 0px;
  opacity: 0.7;
  min-width: 318px;
  max-height: 442px;
  overflow: auto;
  transition: 0.5s;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-bottom: 14px;

  &:hover {
    opacity: 1;
  }
`

export const MargedButton = styled(Button)`
  && {
    margin: 16px 6px;
    min-height: 36px;
  }
`

export const SelectField = styled(NativeSelect)`
  min-width: 200px;
  max-width: 280px;
`

export const SelectContainer = styled.div`
  margin: 8px 0px;
`

export const Label = styled(Typography)`
  && {
    ${({ error }) => (error ? 'color: #e53935;' : '')}
  }
`

export const RateContainer = styled(TableRow)`
  &:hover td:nth-child(2) > div {
    position: relative;
    margin-right: 50px;
    margin-left: -50px;
  }
  &:hover td:nth-child(2) > button {
    opacity: 1;
  }
`

export const Rate = styled.div`
  transition: 0.6s;
  z-index: 10;
`

export const DeleteButton = styled(IconButton)`
  && {
    transition: 0.5s;
    position: absolute;
    margin-left: -34px;
    margin-top: -48px;
    opacity: 0;
    z-index: 5;
  }
`
