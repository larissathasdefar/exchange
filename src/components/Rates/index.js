import React, { PureComponent, Fragment } from 'react'
import _ from 'prop-types'
import Typography from '@material-ui/core/Typography'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import Loading from 'ui/Loading'
import {
  Container, MargedButton, SelectField, SelectContainer, Label,
} from './Rates.styles'

const exchangeMoney = (to, from, rates) => rates['EUR'] / rates[to] * rates[from]

class Rates extends PureComponent {
  state={
    openAddCurrency: false,
    from: '',
    to: '',
    error: false,
  }

  handleToggleOpen = () => {
    this.setState(({ openAddCurrency }) => (
      { openAddCurrency: !openAddCurrency, error: false }
    ))
  }

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value })
  }

  handleAddCurrency = () => {
    const { from, to } = this.state
    const { onAddCurrency } = this.props
    if (from === '' || to === '') {
      return this.setState({ error: true })
    } else {
      onAddCurrency(from, to)
      return this.setState({ error: false, openAddCurrency: false })
    }
  }

  renderAddCurrency = () => {
    const { from, to, error } = this.state
    const { currencies } = this.props
    return <Fragment>
      <SelectContainer>
        <Label error={error && from === '' ? 1 : 0}>
          From: *
        </Label>
        <SelectField
          value={from}
          onChange={this.handleChange}
          name="from">
          <option value="">Select a currency</option>
          { Object.keys(currencies).map(key => (
            <option value={key} key={key}>
              {`${key} - ${currencies[key]}`}
            </option>
          )) }
        </SelectField>
      </SelectContainer>
      <SelectContainer>
        <Label error={error && to === '' ? 1 : 0}>
          To: *
        </Label>
        <SelectField
          value={to}
          onChange={this.handleChange}
          name="to">
          <option value="">Select a currency</option>
          { Object.keys(currencies).map(key => (
            <option value={key} key={key}>
              {`${key} - ${currencies[key]}`}
            </option>
          )) }
        </SelectField>
      </SelectContainer>
      <div>
        <MargedButton
          variant="outlined"
          color="secondary"
          onClick={this.handleToggleOpen}>
          Cancel
        </MargedButton>
        <MargedButton
          variant="contained"
          color="primary"
          onClick={this.handleAddCurrency}>
          Add new
        </MargedButton>
      </div>
    </Fragment>
  }

  render() {
    const { followedRates, rates, currencies } = this.props
    const { openAddCurrency } = this.state
    const hasExchange = currency => rates[currency]
    return (
      <Container elevation={1}>
        <Typography variant="h4" align="center" paragraph>
          Rates
        </Typography>
        {
          Object.keys(rates).length === 0
            ? <Loading />
            : (
              <Table>
                <TableBody>
                  {
                    followedRates.map(({ from, to }) => (
                      <TableRow key={`${from}-${to}`}>
                        <TableCell>
                          <Typography variant="h5">
                            {`1 ${from}`}
                          </Typography>
                        </TableCell>
                        <TableCell align="right">
                          <Typography variant="h5">
                            { hasExchange(to)
                              ? exchangeMoney(from, to, rates).toFixed(2)
                              : 'Loading...' }
                          </Typography>
                          <Typography variant="caption">
                            { currencies[to] }
                          </Typography>
                        </TableCell>
                      </TableRow>
                    ))
                  }
                </TableBody>
              </Table>
            )
        }
        {
          openAddCurrency
            ? this.renderAddCurrency()
            : (
              <MargedButton
                variant="outlined"
                color="primary"
                onClick={this.handleToggleOpen}>
                Add new currency
              </MargedButton>
            )
        }
      </Container>
    )
  }
}

Rates.propTypes = {
  followedRates: _.arrayOf(_.shape({
    form: _.string,
    to: _.string,
  })),
  rates: _.shape({}).isRequired,
  currencies: _.shape({}).isRequired,
  onAddCurrency: _.func.isRequired,
}

export default Rates
