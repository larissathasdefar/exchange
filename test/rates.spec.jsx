import React from 'react'
import { shallow } from 'enzyme'

import Rates from 'components/Rates'

describe('Rates', () => {
  it('Following USD to GBP', () => {
    const component = shallow(
      <Rates
        followedRates={[
          { from: 'USD', to: 'GBP' },
        ]}
        rates={{
          USD: 1.129452,
          GBP: 0.882712,
          EUR: 1
        }}
        currencies={{
          USD: 'United States Dollar',
          GBP: 'British Pound Sterling',
          EUR: 'Euro'
        }}
        onAddCurrency={jest.fn()}
        onRemoveCurrency={jest.fn()}
      />
    )
    expect(component.debug()).toMatchSnapshot()
  })
})