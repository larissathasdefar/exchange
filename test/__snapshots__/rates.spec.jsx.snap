// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rates Following USD to GBP 1`] = `
"<Styled(WithStyles(Paper)) elevation={1}>
  <WithStyles(Typography) variant=\\"h4\\" align=\\"center\\">
    Rates
  </WithStyles(Typography)>
  <WithStyles(Typography) variant=\\"caption\\" align=\\"center\\" paragraph={true}>
    Keep track of your favorites exchanges.
  </WithStyles(Typography)>
  <WithStyles(Table)>
    <WithStyles(TableBody)>
      <Styled(WithStyles(TableRow))>
        <WithStyles(TableCell)>
          <WithStyles(Typography) variant=\\"h5\\">
            1 USD
          </WithStyles(Typography)>
        </WithStyles(TableCell)>
        <WithStyles(TableCell) align=\\"right\\">
          <styled.div>
            <WithStyles(Typography) variant=\\"h5\\">
              0.78
            </WithStyles(Typography)>
            <WithStyles(Typography) variant=\\"caption\\">
              British Pound Sterling
            </WithStyles(Typography)>
          </styled.div>
          <Styled(WithStyles(IconButton)) onClick={[Function: onClick]}>
            <pure(DeleteIcon) fontSize=\\"small\\" />
          </Styled(WithStyles(IconButton))>
        </WithStyles(TableCell)>
      </Styled(WithStyles(TableRow))>
    </WithStyles(TableBody)>
  </WithStyles(Table)>
  <Styled(WithStyles(Button)) variant=\\"outlined\\" color=\\"primary\\" onClick={[Function]}>
    Add new currency
  </Styled(WithStyles(Button))>
</Styled(WithStyles(Paper))>"
`;
